### ISSUE TITLE
Android - Dependency version conflicts - Firebase Analytics, and possibility others

### ISSUE THREAD

**Issue by stopiccot** [OneSignal/OneSignal-Unity-SDK#356]
Why [native android setup](https://documentation.onesignal.com/docs/android-sdk-setup) has custom step to add `onesignal-gradle-plugin` to gradle file but [unity android setup](https://documentation.onesignal.com/docs/unity-sdk-setup) has not?

**Comment by jkasten2** [comment]
@stopiccot That's a good question. [OneSignal-Gradle-Plugin](https://github.com/OneSignal/OneSignal-Gradle-Plugin) mainly is helpful to resolve dependency conflicts with Gradle does not handle well. [Mostly around mixing of incompatible Firebase libraries.](https://github.com/OneSignal/OneSignal-Gradle-Plugin/blob/0.12.10/src/main/groovy/com/onesignal/androidsdk/GradleProjectPlugin.groovy#L153-L165)

With Unity however Google has created [EDM4U ("External Dependency Manager for Unity")](https://github.com/googlesamples/unity-jar-resolver) which we have found for OneSignal dependencies does a better job at resolving conflicting versions.

**Comment by stopiccot** [comment]
But EDM4U under the hood uses gradle - it patches gradle template files that gradle than uses to resolve dependencies. And the only reason I created this issue is that I'm having conflicting firebase library version in my Unity project that has both OneSignal (2.13.4) and Firebase Analytics (7.0.2) sdks added to it.

**Comment by stopiccot** [comment]
And correct if I'm wrong previously Unity Android setup had onesignal-gradle-plugin step. But for some reason it was removed.

**Comment by jkasten2** [comment]
> But EDM4U under the hood uses gradle - it patches gradle template files that gradle than uses to resolve dependencies.

It's possible things have changed or it has tweaked Gradle settings but I remember it giving different results that Gradle alone. Also EDM4U my still have a setting that downloads .aar files in to your `/Assets/Plugin/Android` which may also affect the versions used.

> I'm having conflicting firebase library version in my Unity project that has both OneSignal (2.13.4) and Firebase Analytics (7.0.2) sdks added to it.

Could you provide more details on this? Do you get a specific build error or runtime error?

> And correct if I'm wrong previously Unity Android setup had onesignal-gradle-plugin step. But for some reason it was removed.

Correct, we remove it since it seemed EDM4U was handling things correctly. [OneSignal-Gradle-Plugin](https://github.com/OneSignal/OneSignal-Gradle-Plugin) can still be added to your Unity project if you choose.

**Comment by stopiccot** [comment]
> Also EDM4U may still have a setting that downloads .aar files in to your /Assets/Plugin/Android which may also affect the versions used.

We do not use aar downloading but rather "Patch mainTemplate.gradle" option instead

> Could you provide more details on this? Do you get a specific build error or runtime error?

It's a runtime error 

```
04-09 15:30:26.534: E/OneSignal(32321): Unknown error getting FCM Token
04-09 15:30:26.534: E/OneSignal(32321): java.lang.NoClassDefFoundError: Failed resolution of: Lcom/google/android/gms/common/util/zzx;
04-09 15:30:26.534: E/OneSignal(32321): 	at com.google.firebase.iid.zzaa.<init>(Unknown Source:43)
04-09 15:30:26.534: E/OneSignal(32321): 	at com.google.firebase.iid.zzaa.<init>(Unknown Source:2)
04-09 15:30:26.534: E/OneSignal(32321): 	at com.google.firebase.iid.FirebaseInstanceId.getInstance(Unknown Source:31)
04-09 15:30:26.534: E/OneSignal(32321): 	at com.onesignal.PushRegistratorFCM.getToken(PushRegistratorFCM.java:94)
04-09 15:30:26.534: E/OneSignal(32321): 	at com.onesignal.PushRegistratorAbstractGoogle.attemptRegistration(PushRegistratorAbstractGoogle.java:97)
04-09 15:30:26.534: E/OneSignal(32321): 	at com.onesignal.PushRegistratorAbstractGoogle.access$100(PushRegistratorAbstractGoogle.java:37)
04-09 15:30:26.534: E/OneSignal(32321): 	at com.onesignal.PushRegistratorAbstractGoogle$1.run(PushRegistratorAbstractGoogle.java:84)
04-09 15:30:26.534: E/OneSignal(32321): 	at java.lang.Thread.run(Thread.java:784)
```

We currently workarounded this bug by editing OneSignalDependencies.xml. We changed `com.google.firebase:firebase-messaging` version to `21.0.0` so it matches other Firebase libraries but I think that's exactly what one-signal-gradle plugin is intended to do.

> can still be added to your Unity project if you choose.

Yep, but probably that should be mentioned in OneSignal Unity documentation. I don't think we are the only company in the world to use both FirebaseAnalytics and OneSignal in our projects.

**Comment by stopiccot** [comment]
@jkasten2 Any update on this?

**Comment by jkasten2** [comment]
@stopiccot Thanks for the fullstack trace, the issue is due to a mismatched version of Firebase iid and Firebase Messaging libraries indicated from this part of the stack trace:
```
04-09 15:30:26.534: E/OneSignal(32321): java.lang.NoClassDefFoundError: Failed resolution of: Lcom/google/android/gms/common/util/zzx;
04-09 15:30:26.534: E/OneSignal(32321): 	at com.google.firebase.iid.zzaa.<init>(Unknown Source:43)
04-09 15:30:26.534: E/OneSignal(32321): 	at com.google.firebase.iid.zzaa.<init>(Unknown Source:2)
04-09 15:30:26.534: E/OneSignal(32321): 	at com.google.firebase.iid.FirebaseInstanceId.getInstance(Unknown Source:31)
```

The good news is that Firebase Messaging 22.0.0 no longer depends on the Firebase iid library so this specific version match won't happen when using the latest versions for Firebase release in May 2021.
   - OneSignal-Unity-SDK 2.13.5 and older is not compilable with Firebase Messaging 22.0.0 but there will be OneSignal update shortly.

The bad new of course is not everyone will be using the latest Firebase in Unity so we can document using our OneSignal Gradle Plugin.  There might be a better way this could be detected to instruct users to update Firebase automatically so they don't end up finding out at runtime. However we can start my documenting this.

**Comment by tanaynigam** [comment]
The issue has been resolved on OneSignal-Unity-SDK 2.13.6 . You can download the release from our OneSignal-Unity-SDK release page for version [2.13.6](https://github.com/OneSignal/OneSignal-Unity-SDK/releases/tag/2.13.6) or directly from the [Unity Asset Store](https://assetstore.unity.com/packages/add-ons/services/billing/onesignal-sdk-193316).

We have tested OneSignal-Unity-SDK alongside the latest versions of Firebase Analytics and Google Mobile Ads and were able to verify that OneSignal-Unity-SDK was running correctly upon building it to an Android device.
The environment used for the test are:
* Unity 2020.3.6f1
* Firebase Analytics Unity 8.0.0
* Google Mobile Ads 6.0.1
* OneSignal-Unity-SDK 2.13.6
